File: src/others/3-longest-substring/cpp/src/longest_substring_without_repeating.cpp
#include "longest_substring_without_repeating.hpp"

#include <unordered_set>
#include <algorithm>

using namespace std;
int lengthOfLongestSubstring(string s)
{
    int left = 0;
    int maxLength = 0;
    unordered_set<char> charSet; // Current window's unique characters
    for (int right = 0; right < s.length(); right++)
    {
        // Shrink the window until the duplicate s[right] is removed
        while (charSet.find(s[right]) != charSet.end())
        {
            charSet.erase(s[left]);
            left++;
        }
        charSet.insert(s[right]);
        // Update the best answer for the current window size
        maxLength = max(maxLength, right - left + 1);
    }
    return maxLength;
}

File: src/others/3-longest-substring/typescript/longest_substring_without_repeating.ts
/**
 * Returns the longest length of a substring with all unique characters.
 *
 * Implements the sliding window pattern: expand to the right while tracking the
 * unique characters in a Set, and shift the left edge whenever a duplicate is
 * encountered to keep the window valid.
 */
export function lengthOfLongestSubstring(s: string): number {
  let left = 0;
  let maxLength = 0;
  const charSet = new Set(); // Stores characters currently in the window
  for (let right = 0; right < s.length; right++) {
    // Remove chars from the left until s[right] is unique again
    while (charSet.has(s[right])) {
      charSet.delete(s[left]);
      left++;
    }
    charSet.add(s[right]);
    maxLength = Math.max(maxLength, right - left + 1);
  }
  return maxLength;
}
